java/connector-node/risingwave-sink-es-7/src/main/java/com/risingwave/connector/EsSink.java

```diff
@@ -62,14 +64,83 @@
  */
 public class EsSink extends SinkWriterBase {
     private static final Logger LOG = LoggerFactory.getLogger(EsSink.class);
-    private static final String ERROR_REPORT_TEMPLATE = "Error when exec %s, message %s";
+    private static final String ERROR_REPORT_TEMPLATE = "Error message %s";
 
     private final EsSinkConfig config;
-    private final BulkProcessor bulkProcessor;
+    private BulkProcessor bulkProcessor;
     private final RestHighLevelClient client;
+
+    // Used to handle the return message of ES and throw errors
+    private final RequestTracker requestTracker;
+
     // For bulk listener
     private final List<Integer> primaryKeyIndexes;
 
+    class RequestTracker {
+        // Used to save the return results of es asynchronous writes. The capacity is Integer.Max
+        private final BlockingQueue<EsWriteResultResp> blockingQueue = new LinkedBlockingQueue<>();
+
+        // Count of write tasks in progress
+        private int taskCount = 0;
+
+        void add_err_result(String errorMsg) {
+            blockingQueue.add(new EsWriteResultResp(errorMsg));
+        }
+
+        void add_ok_result(int numberOfActions) {
+            blockingQueue.add(new EsWriteResultResp(numberOfActions));
+        }
+
+        void addWriteTask() {
```

Comment:
When we add new write task, we can clear the items already in the queue to early see the error and throw the exception.